### DISCUSSION TITLE
Capturing from sound file

### DISCUSSION THREAD

**Discussion by franga2000** [ggerganov/ggwave#38] [Q&A] [answered]
Hi! I've been trying to capture ggwave data from a recorded sound file and I was hoping someone else has tried before and might be able to help. I'm using Python, although code examples in any language would be appreciated.

I'm basing my code on the [`receive.py`](https://github.com/ggerganov/ggwave/blob/master/examples/ggwave-py/receive.py) example, which I've confirmed works for my use case by playing my sound clips through a loopback device. I'm using ffmpeg to decode the files (since I'm expecting many formats) and I've tried to match the parameters as closely as I can: 48000Hz sample rate, single channel, 32-bit float samples. The example reads 1024 frames, which in mono 32-bit sound should be `1024*4=4096` bytes, which is the chunk size I'm reading from ffmpeg in.

Does this make sense so far, or have I missed any parameters? The one thing I'm not sure is whether it should be little- or big-endian (is it platform-specific, or does PortAudio have a preference?), but I've tested both and never got any reads.

<details>
<summary>
Here's my code so far, if anyone can spot something obviously wrong (`ggwave_from_file` is the entrypoint)
</summary>

```python
import ggwave
import pyaudio
import ffmpeg

# FRAME_SIZE=4  # 32-bit samples @ 1 channel
FRAME_SIZE = 4
BUFFER_SIZE = 1024*FRAME_SIZE

parameters = ggwave.getDefaultParameters()
parameters["payloadLength"] = 16
instance = ggwave.init(parameters)

def read_audio(filename):
	process = (ffmpeg
		.input(filename)
		.output('-', format='f32le', acodec='pcm_f32le', ar=4800, ac=1)
		.global_args('-map_metadata', '-1', '-vn')
		.overwrite_output()
		.run_async(pipe_stdout=True)
	)

	while process.poll() is None:
		packet = process.stdout.read(BUFFER_SIZE)
		yield packet
	process.wait()


def ggwave_decode(audio):
	for chunk in audio:
		res = ggwave.decode(instance, chunk)
		if res:
			text = res.decode("utf-8")
			yield text


def ggwave_from_file(filename):
	audio = read_audio(filename)
	decoder = ggwave_decode(audio)

	for msg in decoder:
		print(msg)
	
```

</details>

**Comment by franga2000** [comment, marked as answer]
Nevermind, I'm a complete idiot :facepalm:  I had the sample rate set to `4800` instead of `48000`! How this single zero managed to evade me for many hours of debugging is absolutely beyond me.....

Anyways, I guess I'll leave this here as a code example since after fixing the stupid mistake, the code works very well.

**Reply by AngryApples** [reply]
Sorry for the novice question.. but once you've updated the source, how are you re-compiling and installing the updated Python module? I am able to rebuild all of the C++ files/examples/etc. and see them operating with the changes, but what about the Python side of things?
Thanks!!

**Reply by ggerganov** [reply]
Try following these instructions:

https://github.com/ggerganov/ggwave/tree/master/bindings/python

**Reply by AngryApples** [reply]
Much appreciated! Apologies again for the novice-level question.

**Reply by ggerganov** [reply]
No worries - all questions are welcome!
